### PR TITLE
APC PDPM277H: the wrong calculation of total 3ph power for the module…

### PR DESCRIPTION
**Pull request by spezialist1**
… with IEC309-5W wires

If the module with IEC309-5W wires is inserted into the modular PDU APC PDPM277H, total 3ph power is calculated incorrectly as OID2 and OID3 thus are equal "-1". For example, snmpwalk for total 3ph power:

# snmpwalk -v2c -cpublic 192.168.0.16 .1.3.6.1.4.1.318.1.1.22.2.6.1.20
SNMPv2-SMI::enterprises.318.1.1.22.2.6.1.20.1.1 = INTEGER: 17 <-- total 3ph power is 1.7kW for module #1 with IEC309-5W
SNMPv2-SMI::enterprises.318.1.1.22.2.6.1.20.1.2 = INTEGER: -1 <-- we need ignore this value for module #1 with IEC309-5W
SNMPv2-SMI::enterprises.318.1.1.22.2.6.1.20.1.3 = INTEGER: -1 <-- we need ignore this value for module #1 with IEC309-5W

SNMPv2-SMI::enterprises.318.1.1.22.2.6.1.20.2.1 = INTEGER: 9 <-- total 3ph power is 0.9kW for module #2 with IEC309-5W
SNMPv2-SMI::enterprises.318.1.1.22.2.6.1.20.2.2 = INTEGER: -1 <-- we need ignore this value for module #2 with IEC309-5W
SNMPv2-SMI::enterprises.318.1.1.22.2.6.1.20.2.3 = INTEGER: -1 <-- we need ignore this value for module #2 with IEC309-5W

SNMPv2-SMI::enterprises.318.1.1.22.2.6.1.20.9.1 = INTEGER: 2 <-- phase L1 power is 0.2kW for module #9 with IEC309-3W
SNMPv2-SMI::enterprises.318.1.1.22.2.6.1.20.9.2 = INTEGER: 8 <-- phase L2 power is 0.8kW for module #9 with IEC309-3W
SNMPv2-SMI::enterprises.318.1.1.22.2.6.1.20.9.3 = INTEGER: 2 <-- phase L3 power is 0.2kW for module #9 with IEC309-3W

Apparently above, for the module #1 total 3ph power was calculated as 17-1-1=15 (1.5kW) that is wrong. As for IEC309-5W wires modules full capacity has to be calculated, for example, as 17+0+0=17 (1.7kW), and for IEC309-3W wires modules, for example, as 2+8+2=12 (1.2kW).

### CODE REVIEW

**Comment by samilliken** [comment]
Actually, it looks more like you've configured your template incorrectly.

OID1 should be SNMPv2-SMI::enterprises.318.1.1.22.2.6.1.20.1.1
OID2 should be SNMPv2-SMI::enterprises.318.1.1.22.2.6.1.20.2.1
OID3 should be SNMPv2-SMI::enterprises.318.1.1.22.2.6.1.20.3.1

And the processing profile you choose should be Combine3OIDWatts

And the multiplier should be 10.

If you would rather treat each module separately, then you'd need 3
different templates that only have 1 OID in them.

In other words, you shouldn't be calling OIDs that return garbage data to
begin with.

Scott

**Comment by spezialist1** [comment]
Scott, it is possible to insert into APC PDPM277H to 24 different 3ph modules which can have wires as IEC309-5W and 3 x IEC309-3W. Thus in openDCIM will have to create 48 different CDU templates: 24 CDU templates for modules with IEC309-5W and 24 CDU templates for modules with 3 x IEC309-3W.

For example, CDU template for module no.1 with IEC309-5W:
OID1=enterprises.318.1.1.22.2.6.1.20.1.1
Multiplier=0.01
SingleOIDWatts

Example CDU template for module no.1 with 3 x IEC309-3W:
OID1=enterprises.318.1.1.22.2.6.1.20.1.1
OID2=enterprises.318.1.1.22.2.6.1.20.1.2
OID3=enterprises.318.1.1.22.2.6.1.20.1.3
Multiplier=0.01
Combine3OIDWatts

Example CDU template for module no.9 with IEC309-5W:
OID1=enterprises.318.1.1.22.2.6.1.20.9.1
Multiplier=0.01
SingleOIDWatts

Example CDU template for module no.9 with 3 x IEC309-3W:
OID1=enterprises.318.1.1.22.2.6.1.20.9.1
OID2=enterprises.318.1.1.22.2.6.1.20.9.2
OID3=enterprises.318.1.1.22.2.6.1.20.9.3
Multiplier=0.01
Combine3OIDWatts

Scott, as it is better: to correct a php-script to eliminate negative values and create 24 universal CDU templates (for IEC309-5W and 3 x IEC309-3W) or to create 48 CDU templates???

P.S.: Scott, Multiplier shall be 0.01, but not 10. I already about it wrote on http://list.opendcim.org/private/discussion/2015-January/001956.html . In a php-script power.inc.php of OID values divide, but aren't multiplied on a Multiplier.
And OID values of power which are pull from APC PDPM277H have values of the tenth kW. Therefore to correct calculate value of power in Watt, pulled OID values need to be multiplied on 100, or in openDCIM power.inc.php to divide on 0.01 (see my pull requests 597-599).

Very thanks for the support.
